### PR TITLE
Stucking RobustConnection

### PR DESCRIPTION
**Pull request by mosquito**
Fixes stuck RobustConnectionwhen a queue has been declared with the exclusive flag.

### CODE REVIEW

File: aio_pika/robust_connection.py [hunk 1]
```diff
@@ -134,9 +134,18 @@ def channel(self, channel_number: int = None,
         for number, channel in self._channels.items():
             try:
                 await channel.on_reconnect(self, number)
-            except CONNECTION_EXCEPTIONS:
+            except CONNECTION_EXCEPTIONS as e:
                 log.exception('Open channel failure')
-                await self.close()
+
+                if self.connection is not None:
+                    await asyncio.gather(
```
Comment (decaz) [inline aio_pika/robust_connection.py]: Why `gather` but not just `self.close(e)`?

File: aio_pika/robust_connection.py [hunk 2]
```diff
@@ -134,9 +134,18 @@ def channel(self, channel_number: int = None,
...
+                        self.connection.close(e), return_exceptions=True
+                    )
+                    self.connection = None
+
+                # The channel has not been opened successfully.
+                # Staying positive and just retry later.
+                await asyncio.sleep(self.reconnect_interval)
```
Comment (decaz) [inline aio_pika/robust_connection.py]: I guess this has no effect.. Maybe we should get rid of `_on_reconnect` method and rewrite `reconnect` method as follows:
```python
    async def reconnect(self):
        if self.is_closed:
            return

        while True:
            try:
                await super().connect()
            except CONNECTION_EXCEPTIONS:
                log.exception('Connection attempt error')
                await asyncio.sleep(self.reconnect_interval)
            else:
                try:
                    for number, channel in self._channels.items():
                        await channel.on_reconnect(self, number)
                except CONNECTION_EXCEPTIONS as e:
                    log.exception('Open channel failure')
                    if self.connection is not None:
                        await self.close(e)
                        self.connection = None
                else:
                    break
        self._reconnect_callbacks(self)
```
What do you think?

File: aio_pika/robust_connection.py (hunk 1 above)
Comment (mosquito) [reply aio_pika/robust_connection.py]: `self.close()` will set `self.__closed` flag. And connection couldn't be reopened after.

Comment (decaz) [reply aio_pika/robust_connection.py]: Ah, understand.

File: aio_pika/robust_connection.py (hunk 2 above)
Comment (mosquito) [reply aio_pika/robust_connection.py]: I try this, it's broken. I don't want to refactor this part now, just fix production issue.

Comment (decaz) [reply aio_pika/robust_connection.py]: OK, but what this sleep gives here? I think the bug is `await self.close()` which prevents `reconnect` method to connect again (because of `self._closed` flag which you mentioned earlier).

Comment (decaz) [reply aio_pika/robust_connection.py]: If you meant waiting before reconnect attempt then there is already sleep at `_on_connection_close` callback..

Comment (mosquito) [reply aio_pika/robust_connection.py]: Yeah. You are right. After rewrite looks better.

File: aio_pika/pool.py [hunk 3]
```diff
@@ -20,18 +20,19 @@ class Pool:
...
         self.loop = loop or asyncio.get_event_loop()
         self.__max_size = max_size
-        self.__items = asyncio.Queue(loop=self.loop)
+        self.__items = asyncio.Queue()
         self.__constructor = constructor
         self.__constructor_args = args or ()
         self.__created = 0
-        self.__lock = asyncio.Lock(loop=self.loop)
+        self.__lock = asyncio.Lock()
+        self.__close_lock = asyncio.Lock()
```
Comment (decaz) [inline aio_pika/pool.py]: I guess this is for #269. Maybe remove within current PR?

File: aio_pika/robust_connection.py [hunk 4]
```diff
@@ -106,15 +106,25 @@ def add_reconnect_callback(self, callback: Callable[[], None]):
...
+                await channel.on_reconnect(self, number)
+
+            self._reconnect_callbacks(self)
+
+        except CONNECTION_EXCEPTIONS as e:
             log.exception('Connection attempt error')
 
+            if self.connection is not None:
+                await asyncio.gather(
+                    self.connection.close(e), return_exceptions=True
```
Comment (decaz) [inline aio_pika/robust_connection.py]: If you use `return_exceptions=True` to avoid exceptions raised maybe should use following:
```python
with suppress(CONNECTION_EXCEPTIONS):
    await self.connection.close(e)
```

File: aio_pika/pool.py (hunk 3 above)
Comment (mosquito) [reply aio_pika/pool.py]: Nope, commits from this PR will be removed from #269

Comment (decaz) [reply aio_pika/pool.py]: I mean this attribute is not used.

File: aio_pika/robust_connection.py [hunk 5]
```diff
@@ -102,19 +103,36 @@ def add_reconnect_callback(self, callback: Callable[[], None]):
...
         if self.is_closed:
-            return
-
-        try:
-            await super().connect()
-        except CONNECTION_EXCEPTIONS:
-            log.exception('Connection attempt error')
+            raise RuntimeError("{!r} connection closed".format(self))
+
+        async with self._reconnect_lock:
```
Comment (decaz) [inline aio_pika/robust_connection.py]: Maybe we should check the lock before?
```python
if self._reconnect_lock.locked():
    log.warning('Already reconnecting')
    return

async with self._reconnect_lock:
    ...
```

File: aio_pika/robust_connection.py (hunk 4 above)
Comment (mosquito) [reply aio_pika/robust_connection.py]: gather creates a separate task from coroutine. I think this better.

File: aio_pika/pool.py (hunk 3 above)
Comment (mosquito) [reply aio_pika/pool.py]: Yes. Done

**Comment by mosquito** [comment]
@decaz @dizballanze all checks done, please review.

File: aio_pika/pool.py (hunk 3 above)
Comment (decaz) [reply aio_pika/pool.py]: `__close_lock` remained in `__slots__`.

File: aio_pika/robust_connection.py [hunk 6]
```diff
@@ -64,6 +74,7 @@ def _on_connection_close(self, connection, closing, *args, **kwargs):
 
         super()._on_connection_close(connection, closing)
 
+        log.info("Connection to %s closed reconnecting...", self)
```
Comment (decaz) [inline aio_pika/robust_connection.py]: Little grammar fix: "Connection to %s **closed. R**econnecting..."

File: aio_pika/robust_connection.py [hunk 7]
```diff
@@ -44,7 +44,13 @@ def __init__(self, url, loop=None, **kwargs):
 
         self.__channels = set()
         self._reconnect_callbacks = CallbackCollection()
+        self._connect_lock = asyncio.Lock()
         self._closed = False
+        self.connected = asyncio.Event()
```
Comment (decaz) [inline aio_pika/robust_connection.py]: What this event for?

Comment (mosquito) [reply aio_pika/robust_connection.py]: Event for external use. Useful for tests right now but I want to document it.

File: aio_pika/robust_connection.py [hunk 8]
```diff
@@ -77,44 +91,61 @@ def add_reconnect_callback(self, callback: Callable[[], None]):
...
-                if self.fail_fast:
-                    raise
-
-                log.warning(
-                    "First connection attempt failed "
-                    "and will be retried after %d seconds",
-                    self.reconnect_interval,
-                    exc_info=True,
-                )
+        if self.reconnecting:
```
Comment (decaz) [inline aio_pika/robust_connection.py]: Why not to return if already reconnecting?

Comment (mosquito) [reply aio_pika/robust_connection.py]: when the user wants to call reconnect explicitly it's just printed log warning.

File: aio_pika/robust_connection.py [hunk 9]
```diff
@@ -77,44 +91,61 @@ def add_reconnect_callback(self, callback: Callable[[], None]):
...
+    async def __cleanup_connection(self, exc):
+        await asyncio.gather(
+            self.connection.close(exc),
+            return_exceptions=True,
+        )
+        self.connection = None
+
     async def connect(self, timeout: TimeoutType = None, **kwargs):
+        if self.is_closed:
+            raise RuntimeError("{!r} connection closed".format(self))
```
Comment (decaz) [inline aio_pika/robust_connection.py]: This method is for connecting, maybe raising exception here is not needed?

Comment (decaz) [reply aio_pika/robust_connection.py]: Or it is should be at `reconnect` method.

File: aio_pika/robust_connection.py [hunk 10]
```diff
@@ -64,6 +74,10 @@ def _on_connection_close(self, connection, closing, *args, **kwargs):
 
         super()._on_connection_close(connection, closing)
 
+        log.info(
+            "Connection to %s closed. Reconnecting after %r seconds.",
```
Comment (decaz) [inline aio_pika/robust_connection.py]: after **%d** seconds

File: aio_pika/robust_connection.py (hunk 9 above)
Comment (mosquito) [reply aio_pika/robust_connection.py]: Nope. It's a case when the user wants to reopen closed connection.

File: aio_pika/robust_connection.py (hunk 10 above)
Comment (mosquito) [reply aio_pika/robust_connection.py]: `%r` returns same result but without problem when non integer passed.

Comment (decaz) [reply aio_pika/robust_connection.py]: It's a little bit confusing that %s, %d and %r are mixed. I mean sometimes it is %d, but sometimes it is %r and so on..

File: aio_pika/robust_connection.py (hunk 9 above)
Comment (decaz) [reply aio_pika/robust_connection.py]: Do you mean that if robust connection was closed then the user can't open it again?

Comment (mosquito) [reply aio_pika/robust_connection.py]: `self.is_closed` will be changed only after `self.close()` called. But never in reconnecting state.

File: aio_pika/robust_connection.py (hunk 10 above)
Comment (mosquito) [reply aio_pika/robust_connection.py]: My rule is "Type `%r` when you log user-defined parameters"

File: aio_pika/robust_connection.py (hunk 9 above)
Comment (decaz) [reply aio_pika/robust_connection.py]: I understand it. I mean following:
```python
connection = await aio_pika.connect_robust(...)
await connection.close()
await connection.connect()  # This would be illegal?
<RuntimeError>
```

Comment (mosquito) [reply aio_pika/robust_connection.py]: Yep illegal

File: aio_pika/robust_connection.py (hunk 10 above)
Comment (decaz) [reply aio_pika/robust_connection.py]: I mean usage of "after %d seconds" below (line 137) but "after %r seconds" here for example. Anyway, let's merge it.

Comment (mosquito) [reply aio_pika/robust_connection.py]: Fine.

File: aio_pika/robust_connection.py [hunk 11]
```diff
@@ -133,10 +133,8 @@ def add_reconnect_callback(self, callback: Callable[[], None]):
                     await self.__cleanup_connection(e)
 
                     log.warning(
-                        "Connection attempt to \"%s\" failed "
-                        "and will be retried after %d seconds",
```
Comment (decaz) [inline aio_pika/robust_connection.py]: Why did you remove this line? `_on_connection_close` doesn't log it while reconnecting..

Comment (mosquito) [reply aio_pika/robust_connection.py]: `self._cleanup_connection` will log same log.

Comment (decaz) [reply aio_pika/robust_connection.py]: You mean here? https://github.com/mosquito/aio-pika/blob/1065083434c4ef3b45c653e30bc78e8637b8dad1/aio_pika/robust_connection.py#L77-L80
But it won't log while reconnecting due to https://github.com/mosquito/aio-pika/blob/1065083434c4ef3b45c653e30bc78e8637b8dad1/aio_pika/robust_connection.py#L63-L65

Comment (mosquito) [reply aio_pika/robust_connection.py]: Do you want to swap it?

Comment (mosquito) [reply aio_pika/robust_connection.py]: `_on_connection_close` will be called when every connection closed. In [this case](https://github.com/mosquito/aio-pika/blob/1065083434c4ef3b45c653e30bc78e8637b8dad1/aio_pika/robust_connection.py#L63-L65) connection is establishing and this condition needed only for `self.__cleanup_connection`.

Comment (decaz) [reply aio_pika/robust_connection.py]: I thought you want to log about reconnection interval here https://github.com/mosquito/aio-pika/blob/1065083434c4ef3b45c653e30bc78e8637b8dad1/aio_pika/robust_connection.py#L77-L80
OK, but now after removal of lines about it I don't see how `self._cleanup_connection` would log it, - there is no any logging in it.

Comment (mosquito) [reply aio_pika/robust_connection.py]: What do you propose? A simple way is copying log from [here](https://github.com/mosquito/aio-pika/blob/1065083434c4ef3b45c653e30bc78e8637b8dad1/aio_pika/robust_connection.py#L77-L80) to `connect` function I guess.

Comment (decaz) [reply aio_pika/robust_connection.py]: I suggest to save _"and will be retried after %d seconds"_ part of log within `connect` method. I don't see how it will be logged twice.

File: aio_pika/robust_connection.py [hunk 12]
```diff
@@ -77,44 +91,61 @@ def add_reconnect_callback(self, callback: Callable[[], None]):
...
+
+                    await self.__cleanup_connection(e)
+
+                    log.warning(
+                        "Connection attempt to \"%s\" failed. "
+                        "Reconnecting after %r seconds.",
+                        self, self.reconnect_interval, exc_info=True,
+                    )
+                except asyncio.CancelledError as e:
+                    await self.__cleanup_connection(e)
```
Comment (decaz) [inline aio_pika/robust_connection.py]: @mosquito shouldn't we break the loop here?
